source/code/plugin/in_containerinventory.rb

```diff
@@ -0,0 +1,234 @@
+          if !portBindings.nil?
+            portBindingsString = portBindings.to_s
+            instance['Ports'] = (portBindingsString == "null")? "" : portBindingsString
+          end
+        else
+          $log.info("Attempt in ObtainContainerHostConfig to get container: #{container['Id']} host config information returned null")
+        end
+        rescue => errorStr
+          $log.warn("Exception in obtainContainerHostConfig: #{errorStr}")
+        end
+    end
+
+    def inspectContainer(id, nameMap)
+      containerInstance = {}
+      begin
+        container = DockerApiClient.dockerInspectContainer(id)
+        if !container.nil? && !container.empty?
+          containerInstance['InstanceID'] = container['Id']
+          containerInstance['CreatedTime'] = container['Created']
+          containerName = container['Name']
+          if !containerName.nil? && !containerName.empty?
+            # Remove the leading / from the name if it exists (this is an API issue)
+            containerInstance['ElementName'] = (containerName[0] == '/') ? containerName[1..-1] : containerName
+          end
+          imageValue = container['Image']
+          if !imageValue.nil? && !imageValue.empty?
+            containerInstance['ImageId'] = imageValue
+            repoImageTagArray = nameMap[imageValue]
+            if nameMap.has_key? imageValue
+              containerInstance['Repository'] = repoImageTagArray[0]
+              containerInstance['Image'] = repoImageTagArray[1]
+              containerInstance['ImageTag'] = repoImageTagArray[2]
+            end
+          end
+          obtainContainerConfig(containerInstance, container);
+          obtainContainerState(containerInstance, container);
+          obtainContainerHostConfig(containerInstance, container);
+        end
+      rescue => errorStr
+          $log.warn("Exception in inspectContainer: #{errorStr} for container: #{id}")
+      end
+      return containerInstance
+    end
+
+    def enumerate
+      currentTime = Time.now
+      emitTime = currentTime.to_f
+      batchTime = currentTime.utc.iso8601
+      containerInventory = Array.new
+      $log.info("in_container_inventory::enumerate : Begin processing @ #{Time.now.utc.iso8601}")
+      hostname = DockerApiClient.getDockerHostName
+      begin
+        containerIds = DockerApiClient.listContainers
+        if !containerIds.empty?
+          eventStream = MultiEventStream.new
+          nameMap = DockerApiClient.getImageIdMap
+          containerIds.each do |containerId|
+            inspectedContainer = {}
+            inspectedContainer = inspectContainer(containerId, nameMap)
+            inspectedContainer['Computer'] = hostname
+            containerInventory.push inspectedContainer
+            ContainerInventoryState.writeContainerState(inspectedContainer)
+          end
+          # Update the state for deleted containers
+          deletedContainers = ContainerInventoryState.getDeletedContainers(containerIds)
```

Comment:
Consider moving this to writeContainerState. The reasoning is you are updating the ContainerInventory folder in readContainerState which is essentially a write operation.

Reply:
writeContainerState writes the container information into the disk with this plugin invocation. getDeletedContainers deletes the the container after the state is updated to deleted. I dont think these 2 can be combined. We can discuss offline.